wdl-lint/src/rules/disallowed_declaration_name.rs

```diff
@@ -0,0 +1,295 @@
+    }
+
+    fn bound_decl(&mut self, state: &mut Self::State, reason: VisitReason, decl: &BoundDecl) {
+        if reason == VisitReason::Enter {
+            check_decl_name(state, &Decl::Bound(decl.clone()), &self.exceptable_nodes());
+        }
+    }
+
+    fn unbound_decl(&mut self, state: &mut Self::State, reason: VisitReason, decl: &UnboundDecl) {
+        if reason == VisitReason::Enter {
+            check_decl_name(
+                state,
+                &Decl::Unbound(decl.clone()),
+                &self.exceptable_nodes(),
+            );
+        }
+    }
+}
+
+impl Rule for DisallowedDeclarationNameRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Declaration names should not end with their type"
+    }
+
+    fn explanation(&self) -> &'static str {
+        "Declaration names should not include their type as a suffix. \
+        This makes the code more verbose and often redundant. For example, use \
+        'counter' instead of 'counterInt', or 'is_active' instead of 'activeBool'."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Style, Tag::Clarity])
+    }
+
+    fn exceptable_nodes(&self) -> Option<&'static [SyntaxKind]> {
+        Some(&[
+            SyntaxKind::VersionStatementNode,
+            SyntaxKind::BoundDeclNode,
+            SyntaxKind::UnboundDeclNode,
+            SyntaxKind::InputSectionNode,
+            SyntaxKind::OutputSectionNode,
+        ])
+    }
+}
+
+/// Check declaration name for type suffixes only
+fn check_decl_name(
+    state: &mut Diagnostics,
+    decl: &Decl,
+    exceptable_nodes: &Option<&'static [SyntaxKind]>,
+) {
+    let name = decl.name();
+    let name_str = name.as_str();
+    let name_lower = name_str.to_lowercase();
+    
+    // Get the declaration type
+    let ty = decl.ty();
+    
+    // Skip type reference types (user-defined structs)
+    if ty.as_type_ref().is_some() {
+        return;
+    }
+    
+    // Extract type names to check
+    let mut type_names = Vec::new();
+    
+    // Add primitive type if present
+    if let Some(primitive_type) = ty.as_primitive_type() {
+        // Skip File and String types as they cause too many false positives
+        match primitive_type.kind() {
+            PrimitiveTypeKind::File | PrimitiveTypeKind::String => {},
+            PrimitiveTypeKind::Boolean => {
+                // For Boolean, check both "Boolean" and "Bool"
+                type_names.push("Boolean".to_string());
+                type_names.push("Bool".to_string());
+            },
+            _ => type_names.push(primitive_type.to_string()),
+        }
+    }
+    
+    // Add compound types
+    if ty.as_array_type().is_some() {
+        type_names.push("Array".to_string());
+    } else if ty.as_map_type().is_some() {
+        type_names.push("Map".to_string());
+    } else if ty.as_pair_type().is_some() {
+        type_names.push("Pair".to_string());
+    }
```

Comment:
This would be better if it didn't hardcode these strings, as they should be derivable from something in `wdl-ast` (although offhand, I can't remember how exactly).

What do you think about the inner types? For example, should `Array[Float] my_float` trigger a lint? That could be made trickier by plurals (e.g. `Pair[Float, Float] my_floats`)